tfrecord/reader.py

```diff
@@ -69,14 +69,16 @@ def read_records(start_offset=None, end_offset=None):
     file.close()
 
 
-def tfrecord_loader(data_path, index_path, description, shard=None):
-    """Create an iterator from a tfrecord dataset. 
+def tfrecord_loader(data_path, index_path, description=None, shard=None):
+    """Create an iterator from a tfrecord dataset.
 
     Args:
         data_path: Path of the input data.
         index_path: Path of index file. This can be set to None if not available.
-        description: A dictionary of key and values where keys are the name of the features and values correspond to
-                     data type. The data type can be "byte", "float" or "int".
+        description: A dictionary of (key, value) pairs where the keys are the
+            name of the features and values (ignored) correspond to data type.
+            Only the specified keys are extracted. If None, then all features
+            contained in the file are extracted. (default: None)
```

Comment:
I suggest to change this in this way: Description can be either a list of key names or a dictionary of name: type. If the type is specified it should be as before "byte", "float", "int" and we should use the inferred type to verify it's compatible with user provided type.

Reply:
There are definitely pros and cons to each approach. Using the list approach makes it easier to just define which specific keys the user wants to extract. If None (default), then all the keys are extracted. However, this will sacrifice backward compatibility with previous versions as the user will have to pass in a list rather than a dict. Obviously, this can be easily solved some sort of `DeprecationWarning`.

If we go with the type-specification approach, then that will add unnecessary complexity with verifying that the user has specified the right type. For example, if the user hasn't specified a compatible type for the specific key, then we have to return a `TypeError` exception. I think if the user intention is to load a specific dtype into another compatible one using this approach, then the user can do that independently later using `torch.type(dtype)`.

I prefer using the list approach as its easier, and we don't have to worry about the type checking/compatibility.

Reply:
That's not what I said. Description should either be None, a list, or a dictionary. If a dictionary the values should not be ignored it should be used to verify that what user expects is there. Please read my comment again.